File: Python/0153-m-find-minimum-in-rotated-sorted-array.py
"""
Binary Search
Time: O(logn)
Space: O(1)
"""
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if not nums:
            return -1
        
        start, end = 0, len(nums) - 1
        target = nums[end]
        while start + 1 < end:
            mid = start + (end - start) // 2
            if nums[mid] <= target:
                end = mid
            else:
                start = mid
        
        if nums[start] <= target:
            return nums[start]
        else:
            return nums[end]

File: leetcode/C++/0153-m-find-minimum-in-rotated-sorted-array.cpp
// Binary search
// Time: O(logN)
// Space: O(1)
class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        while (l < r - 1) {
            if (nums[l] < nums[r])
                return nums[l];
            int mid = l + (r - l) / 2;
            if (nums[mid] > nums[l])
                l = mid;
            else
                r = mid;
        }
        return min(nums[l], nums[r]);
    }
};